### PR TITLE
Fixed EmailValidator to check email length properly

### LINKED ISSUE #10142: Validator email Yii2
class User extends \yii\base\Model 
{
...
 public function rules() 
    {
        return [
        ['email', 'email', 'message' => 'Invalid email'],
];
....
}

$model->validate() do not work with email max length. It may be 256 symbols. But when i input more then 256 symbols, it return, then email is valid. 
P.S. Sorry, i am russian =)

### PR DESCRIPTION
**Pull request by SilverFire**
Closes #10142
- [x] PHP & JS
- [x] Tests
- [x] Changelog
- [x] Code review

### CODE REVIEW

File: framework/validators/EmailValidator.php [hunk 1]
```diff
@@ -70,19 +70,34 @@ public function init()
...
-                $value = $matches[1] . idn_to_ascii($matches[2]) . '@' . idn_to_ascii($domain) . $matches[4];
+                $matches['local'] = idn_to_ascii($matches['local']);
+                $matches['domain'] = idn_to_ascii($matches['domain']);
+                $value = $matches['name'] . $matches['open'] . $matches['local'] . '@' . $matches['domain'] . $matches['close'];
             }
-            $valid = preg_match($this->pattern, $value) || $this->allowName && preg_match($this->fullPattern, $value);
-            if ($valid && $this->checkDNS) {
-                $valid = checkdnsrr($domain, 'MX') || checkdnsrr($domain, 'A');
+
+            if (strlen($matches['local']) > 64 || mb_strlen($matches['name']) > 64) {
```
Comment (samdark) [inline framework/validators/EmailValidator.php]: The order of checks here looks a bit weird. Also using default encoding for mb_strlen is not a good idea.

Comment (SilverFire) [inline framework/validators/EmailValidator.php]: First we check the length of a local part of email address, then - the length of name. I don't think it's weird.

Could you suggest an idea for `mb_strlen`?

Comment (samdark) [inline framework/validators/EmailValidator.php]: OK. I was wrong about that.

**Comment by samdark** [comment]
Thank you!

**Comment by nkovacs** [comment]
Why is the display name (if allowName is true) not allowed to be more than 64 characters?

The linked rfc (http://tools.ietf.org/html/rfc5321#section-4.5.3.1.1) does not say anything about the display name.

The display-name is specified here: http://tools.ietf.org/html/rfc5322#section-3.4, but there's nothing about lengths in there.

**Comment by samdark** [comment]
https://stackoverflow.com/questions/386294/what-is-the-maximum-length-of-a-valid-email-address/574698#574698

**Comment by nkovacs** [comment]
`John Doe <johndoe@example.org>` is not a valid email address at all according to this, which is linked from that answer: http://isemail.info/John%20Doe%20%3Cjohndoe%40example.org%3E

That's because `John Doe <johndoe@example.org>` is the address format used in the message, in the from field, specified in RFC5322. For SMTP (RFC5321), you have to strip the display name, and only use `johndoe@example.org`. At least that's how I understand it. The php mail function probably does this for you automatically.

**Comment by SilverFire** [comment]
> John Doe <johndoe@example.org> is not a valid email address at all according to this, which is linked from that answer: http://isemail.info/John%20Doe%20%3Cjohndoe%40example.org%3E

This test is not designed to validate the name of recipient, as far as I know.


@nkovacs I didn't quite catch your thought.

> Why is the display name (if allowName is true) not allowed to be more than 64 characters?

Do you think that the local part of the email address must not be limited?

**Comment by nkovacs** [comment]
You're confusing the display name with the local part.

In `John Doe <johndoe@example.org>`, `John Doe` is the display name, `johndoe` is the local part. The local part must not be longer than 64 characters. The display name can be any length, because it's not used in SMTP, it's not even allowed in the path.
It's used in the message body itself, which is governed by RFC5322, and that doesn't specify a length for the display name.

**Comment by nkovacs** [comment]
Try this:

```php
<?php

$email = 'your@email.here'

mail('Test this is a really long display name that is longer than 400 characters Lorem ipsum dolor sit amet consectetur adipiscing elit. Donec magna neque imperdiet et orci a fringilla viverra massa. Aenean ullamcorper fringilla odio eu consequat. Donec euismod semper nunc non venenatis. Sed ac placerat nisi. Quisque tempus purus eu nisl vestibulum tempus facilisis velit molestie. Quisque et varius ligula. Sed ullamcorper diam quis fermentum venenatis ante est cras amet <' . $email  . '>', 'Test', 'test message php');
```

It works for me.

**Comment by SilverFire** [comment]
> You're confusing the display name with the local part.

That was a logical typo, sorry :)

I tend to agree, it seems that you're right.

What about http://tools.ietf.org/html/rfc5321#section-4.5.3.1.1? What do they mean in words "total length of a user name or other local-part"? What kind of "other local-part"?

**Comment by nkovacs** [comment]
Yeah, that wording is a bit confusing.
It says "user name or other local part", so to me that means that the local part can be a user name, but it can also be something else (also supported by http://tools.ietf.org/html/rfc5321#section-2.3.11).

**Comment by SilverFire** [comment]
Fixed in 127eeb5
Thank you @nkovacs

**Pull request merged**